src/Microsoft.AspNet.Authentication/AuthenticationHandler.cs

```diff
@@ -61,12 +64,13 @@ internal AuthenticationOptions BaseOptions
 
         public bool Faulted { get; set; }
 
-        protected async Task BaseInitializeAsync([NotNull] AuthenticationOptions options, [NotNull] HttpContext context, [NotNull] ILogger logger)
+        protected async Task BaseInitializeAsync([NotNull] AuthenticationOptions options, [NotNull] HttpContext context, [NotNull] ILogger logger, [NotNull] IUrlEncoder encoder)
         {
             _baseOptions = options;
             Context = context;
             RequestPathBase = Request.PathBase;
             Logger = logger;
+            UrlEncoder = encoder;
```

Comment:
Part of the trouble is who should add it as a service? Look at my last PR that added this to cookies. Most of the samples had to add the service themselves, which seems inappropriate to require for middleware that don't even use it.  If you can improve the service registration then that will help.